### ISSUE TITLE
How best to submit upgrade of CURL?

### ISSUE THREAD

**Issue by ned14** [cpp-pm/hunter#325]
At work we badly needed a modern curl, so I went and merged latest curl 7.74 into the hunter branch: https://github.com/ned14/curl/tree/hunter-7.74.0

Because latest curl uses latest hunter which comes with OpenSSL 1.1 instead of 0.98, I also had to upgrade libssh2 to latest 1.9.0: https://github.com/ned14/libssh2/tree/hunter-1.9.0

Finally, these then broke libs3, so I had to fix that too: https://github.com/ned14/libs3/tree/hunter-4.1.0-p2

My concern here is that if I update libssh2, existing curl will break unless I use hunter config version pins. So what process ought I follow here? How does one go about submitting upgrades to three libraries at once without doing each individually?

**Comment by rbsheth** [comment]
You can just not upgrade the defaults until the dependencies are ready.

**Comment by ned14** [comment]
> You can just not upgrade the defaults until the dependencies are ready.

Ok, so do I submit PR's for each of libssh2, curl and libs3 first? Once all are accepted and get their own branch, only then do I raise the PR against hunter with updated default versions for all three libraries? Then all three get tested together on CI, and you'll know if my triple set of upgrades are good?

**Comment by rbsheth** [comment]
Let's upgrade each and set the defaults as normal. CI will go through and then we can decide to revert the default before merge to defer it. Sounds good?

**Comment by ned14** [comment]
Be aware that if we upgrade libssh2 and change the default, the existing curl will fail to compile. Equally, the upgraded curl will fail to compile unless we do change the default.

This is why I raised this problem here before I begin, because I think you'll need to accept that cpp-pm will become broken for a short while.

Anyway, I've gone ahead and opened the PRs for all three libraries. I appreciate that reviewing the changes will take a while, so let me know when you're done with all three, then I'll prepare the hunter PR.

**Comment by rbsheth** [comment]
I'm aware of that, but we can just avoid making a release until all the changes have landed. I appreciate the thought that went into the sequencing!

Let's get these changes in and see what happens...

**Comment by ned14** [comment]
I'll just link to each of the PRs here to save people from having to refind them constantly:

- https://github.com/cpp-pm/curl/pull/1
- https://github.com/cpp-pm/libssh2/pull/1
- https://github.com/cpp-pm/libs3/pull/2

**Comment by rbsheth** [comment]
@ned14 Happy new year! Here are all the releases:

https://github.com/cpp-pm/libssh2/releases/tag/v1.9.0-p0
https://github.com/cpp-pm/curl/releases/tag/v7.74.0-p0
https://github.com/cpp-pm/libs3/releases/tag/v4.1.0-287e4be-p1

**Comment by ned14** [comment]
The PR can be found at https://github.com/cpp-pm/hunter/pull/336

hunter-testing appears to no longer be used, so I'm assuming the hunter github actions "do the right thing"?

**Comment by rbsheth** [comment]
> The PR can be found at #336

You need to add the versions to each project's `hunter.cmake` file in that PR.

> hunter-testing appears to no longer be used, so I'm assuming the hunter github actions "do the right thing"?

Correct, in fact batching all updates will allow for them to be tested together.

**Comment by ned14** [comment]
Ok, the remaining CI failures are in CURL in these matrix items:

- Android NDK builds
- All the Windows builds
- Some, but not all, of the Mac OS builds.

... and all are of the form:

```sh
try_compile:
  Cannot copy output executable

    ''

  to destination specified by COPY_FILE:
```
The problem is that the path `try_compile()` is trying to copy to does not exist i.e. it chose a non-existent path. What's really weird is that earlier on in the cmake, `try_compile()` works just fine, but at some later point it stops working.

Note that the Windows builds work perfectly on my machine here.

I'm going to propose that I don't think that these failures are ones I can fix? [If you read the many cases where the same thing happens off Google](https://www.google.com/search?safe=off&biw=1303&bih=1257&ei=bd_1X_zIJtGG1fAPhtWd0AI&q=try_compile+++Cannot+copy+output+executable++++++%27%27++++to+destination+specified+by+COPY_FILE%3A&oq=try_compile+++Cannot+copy+output+executable++++++%27%27++++to+destination+specified+by+COPY_FILE%3A&gs_lcp=CgZwc3ktYWIQA1CcpAFY57gBYJe6AWgAcAJ4AIABU4gB3wWSAQIxMZgBAaABAaoBB2d3cy13aXrAAQE&sclient=psy-ab&ved=0ahUKEwi8s_rD14fuAhVRQxUIHYZqByo4ChDh1QMIDQ&uact=5), it appears to almost always be a mistake in the compiler or toolchain configuration which causes this. The general recommendations tend of be of the forms of to wipe your build directory, use a toolchain file instead of command line cmake variable overrides or environment variables, and generally when people change up how they configure cmake, the problem resolves. This particular failure goes back to cmake 2.8 days, and the cmake developers generally respond "you're using cmake wrong" and insta-close the bug report.

I've examined your github actions CI scripts and nothing leaps out at me as an obvious cause.

Not sure what to do next?

**Comment by rbsheth** [comment]
Weird that the Windows builds work on your machine but don't in CI. Can you check the older version of CURL that we have and see if it has the same problem? You can open a dummy PR.

**Comment by ned14** [comment]
So, judging from https://github.com/cpp-pm/hunter/pull/340, there must be something which either upstream CURL's merged CMakeLists is doing, or a bad merge by me, which is causing the failures in #336. Sigh. It's never easy ...

I'll see if I can carve out some time next week to deep dive on the cause. As I mentioned, it works here just fine, so I expect this to be painful.

**Comment by ned14** [comment]
BTW note https://github.com/cpp-pm/curl/pull/2. No idea if that fixes the problem, but worth a try.

**Comment by rbsheth** [comment]
> BTW note [cpp-pm/curl#2](https://github.com/cpp-pm/curl/pull/2). No idea if that fixes the problem, but worth a try.

@ned14 Did you try that fix in CI?

**Comment by ned14** [comment]
> @ned14 Did you try that fix in CI?

How do I test it in your CI? I don't think my fork runs your CI scripts properly?

**Comment by rbsheth** [comment]
@ned14 If your fork has the latest version of master you should be able to use Github Actions the same way we do here

**Comment by ned14** [comment]
> @ned14 If your fork has the latest version of master you should be able to use Github Actions the same way we do here

Take a look at https://github.com/ned14/hunter/actions/runs/503992028. I rebased that onto tip of your master branch before I pushed.

**Comment by rbsheth** [comment]
> > @ned14 If your fork has the latest version of master you should be able to use Github Actions the same way we do here
> 
> Take a look at https://github.com/ned14/hunter/actions/runs/503992028. I rebased that onto tip of your master branch before I pushed.

It says "Strategy produced more than 256" - probably you are generating too many jobs somehow. Maybe we can try without `libs3`?

**Comment by ned14** [comment]
Weird the same thing works when the PR is targeting hunter trunk right? I'll try removing s3 there now, thanks.

**Comment by ned14** [comment]
Once s3 is removed, failure is now due to:

```
[hunter ** FATAL ERROR **] Upload: PASSWORD is missing
[hunter ** FATAL ERROR **] [Directory:/home/runner/work/hunter/hunter/_testing/Hunter/_Base/Download/Hunter/unknown/d324762/Unpacked/cmake/projects/OpenSSL]
```

**Comment by rbsheth** [comment]
Ugh, can you just open another dummy PR here? Sorry about the annoyance

**Comment by ned14** [comment]
Did not https://github.com/cpp-pm/hunter/pull/336 automatically update with my force push? It looks like it did. Unless you want an additional dummy PR here?

**Comment by rbsheth** [comment]
Yeah I see them. Looks like Mac is now failing with the same message. I will try updating the CMake version used for testing, that might help https://github.com/cpp-pm/polly/blob/master/bin/install-ci-dependencies.py#L196

**Comment by rbsheth** [comment]
https://github.com/cpp-pm/polly/commit/5acd508403b358c5e11149c4e7133f04a4904b30

**Comment by rbsheth** [comment]
@ned14 It looks like that didn't really help. Let's just merge as-is since you verified it works locally for you.
